lib/modules/selectedEntry.js

```diff
@@ -251,88 +218,32 @@ function updateActiveElement(selected, last) {
-let lastSelectedCache;
-
-const setupLastSelectedCache = _.once(async () => {
-	lastSelectedCache = await Session.get(lastSelectedKey) || {};
-
-	// clean cache every so often and delete any urls that haven't been visited recently
-	const clearCachePeriod = 21600000; // 6 hours
-	const itemExpiration = 3600000; // 1 hour
-	const now = Date.now();
-	if (!lastSelectedCache.lastScan || (now - lastSelectedCache.lastScan > clearCachePeriod)) {
-		for (const idx in lastSelectedCache) {
-			if (lastSelectedCache[idx] && (now - lastSelectedCache[idx].updated > itemExpiration)) {
-				delete lastSelectedCache[idx];
-			}
-		}
-		lastSelectedCache.lastScan = now;
-		Session.set(lastSelectedKey, lastSelectedCache);
-	}
-});
-
-function urlForSelectedCache() {
-	let url = document.location.pathname;
-	// remove any trailing slash from the URL
-	if (url.endsWith('/')) {
-		url = url.slice(0, -1);
-	}
-
-	return url;
+	if (selectedThing && elementInViewport(selectedThing.entry)) return;
+	const mostVisible = _.maxBy(Thing.visibleThings(), ({ entry }) => getPercentageVisibleYAxis(entry));
+	if (mostVisible) select(mostVisible);
 }
 
-function updateLastSelectedCache(selected) {
-	if (!lastSelectedCache) return;
+const lastSelectedId = history.state && history.state.lastSelectedId;
 
-	const url = urlForSelectedCache();
-	lastSelectedCache[url] = {
-		fullname: selected.getFullname(),
-		updated: Date.now(),
-	};
-	Session.set('RESmodules.selectedThing.lastSelectedCache', lastSelectedCache);
-}
+function selectInitial(thing) {
+	if (selectedThing) return;
+	if (thing.getFullname() !== lastSelectedId) return;
+	const target = thing.getClosestVisible();
+	if (!target) return;
 
-const selectInitial = _.once(fastAsync(function*(firstThing) {
 	if (module.options.scrollToSelectedThingOnLoad.value) history.scrollRestoration = 'manual';
 	// `scrollRestoration` may also be set in neverEndingReddit
 	const scrollToSelected = history.scrollRestoration === 'manual';
 
-	const lastSelected = yield findLastSelectedThing();
-	const target = (lastSelected || firstThing).getClosestVisible() || lastSelected || firstThing;
-
-	select(target, {
-		scrollStyle: scrollToSelected ? 'legacy' : 'none',
-		paintImmediate: true,
-	});
-
-	addListener(updateLastSelectedCache);
-}));
-
-async function findLastSelectedThing() {
-	if (!module.options.selectLastThingOnLoad.value) {
-		return false;
-	}
-
-	await setupLastSelectedCache();
-
-	const url = urlForSelectedCache();
-	const lastSelected = lastSelectedCache[url] && lastSelectedCache[url].fullname;
-	if (lastSelected) {
-		return Thing.things().find(v => v.getFullname() === lastSelected);
-	}
+	select(target, { scrollStyle: scrollToSelected ? 'legacy' : 'none' });
 }
 
+const addLastSelectedListener = _.once(() => {
+	addListener(selected => {
+		history.replaceState({ ...history.state, lastSelectedId: selected.getFullname() }, document.title);
+	}, 'beforeScroll');
```

Comment:
Should this be done at idle? Or is it too cheap to matter? I vaguely remember someone saying that replaceState/pushState were slow on underpowered machines, but I could be wrong.

Reply:
`replaceState` is unfortunately fairly costly for this relatively unimportant feature, so I switched to using `sessionStorage` instead.